Sources/PackageCollections/PackageCollections+Vallidation.swift

```diff
@@ -0,0 +1,91 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import struct Foundation.NSRange
+import class Foundation.NSRegularExpression
+
+extension PackageCollections {
+    struct Validator {
+        // TODO: configuration?
+        let maximumPackages: Int = 50
+
+        func validate(_ collection: PackageCollectionsModel.Collection) -> [ValidationError]? {
+            var errors: [ValidationError]?
+            let appendError = { (error: ValidationError) in
+                if errors == nil {
+                    errors = .init()
+                }
+                errors?.append(error)
+            }
+
+            if collection.packages.count > self.maximumPackages {
+                appendError(.property(
+                    name: "PackageGroup.packages",
+                    description: "Only \(self.maximumPackages) allowed per group. \(collection.name) has \(collection.packages.count)."
+                ))
+            }
+
+            // Each package
+            // TODO: check repository URL
+            // TODO: requires remote repo?
+
+            // Each package version
+            // TODO: check max num of major versions (2)
+            // TODO: check max num of minor versions (3)
+            // TODO: check versions are in reversed order (requires semver)
+            // copy code from package-feed-tool
+
+            return errors
+        }
+    }
+}
+
+extension PackageCollectionsModel.CollectionSource {
+    static let feedURLPattern = "^https://"
+
+    func validate() -> [ValidationError]? {
```

Comment:
I wonder if this should take a `DiagnosticsEngine` instead, and return a boolean about whether or not it succeeded.  This would allow it to represent warnings, etc, in the future.

Reply:
right now these bubble as errors to the caller which can then feed them to the right channel like the `DiagnosticsEngine`

Reply:
That makes sense.